File: CPP/AbstractClass.cpp
#include<iostream>
#include<cmath>
using namespace std;
//#define PI 3.14
class shape
{
    public:
        const float PI = 3.14;     //Error in java
        virtual void input() = 0;   //pure virtual function
        virtual float area() = 0;

};

class Circle : public shape
{
    float rad;
   public:
       void input()
       {
            cout<<"Enter radius: ";
            cin>>rad;
       }
       float area()
       {
           return PI * rad * rad;
       }
 };

 class Rectangle : public shape
 {
    float l, b;
   public:  
     void input()
     {
        cout<<"\nEnter length: ";
        cin>>l;

        cout<<"Enter breath: ";
        cin>>b;
     }
     float area()
     {
         return l * b;
     }

 };

 class Triangle : public shape
 {
    float a, b, c, s ;
  public:
   void input()
   {
      cout<<"\nEnter the value of a: ";
      cin>>a;

      cout<<"Enter the value of b: ";
      cin>>b;

      cout<<"The the value of c: ";
      cin>>c;
   }
   float area()
   {
      s = (a+b+c)/2;
      return sqrt(s*(s-a) * (s-b) * (s-c));
   }
 };

 int main()
{
   shape *s = new Circle();
   s->input();
   cout<<"The area of circle is "<<s->area()<<endl;

   s = new Rectangle();
   s->input();
   cout<<"The area of rectangle is "<<s->area()<<endl;

   s = new Triangle();
   s->input();
   cout<<"The area of triangle is "<<s->area()<<endl;
}

File: Java/AbstractClass.java
import java.util.Scanner;
abstract class shape {
   public final float PI = 3.14F; 
   public abstract void input();
   public abstract float area();
}

class Circle extends shape {
    private float rad;

    public void input(){
       Scanner sc = new Scanner(System.in); 
       System.out.print("Enter radius: ");
       rad =sc.nextFloat();
      // sc.close();
    }
    public float area(){
        return PI * rad *rad;
    }
}

class Rectangle extends shape {
    private int l,b;

    public void input(){
        Scanner sc = new Scanner(System.in);
        System.out.print("\nEnter length: ");
        l =sc.nextInt();

        System.out.print("Enter breath: ");
        b =sc.nextInt();
        //sc.close();
    }
    public float area(){
        return l * b;
    }
   
}

class Triangle extends shape{
    private int b, h;

    public void input(){
        Scanner sc = new Scanner(System.in);
        System.out.print("\nEnter base: ");
        b =sc.nextInt();

        System.out.print("Enter height: ");
        h =sc.nextInt();
        //sc.close();

    }
    public float area(){
        return (b * h)/2;
    }

} 

class AbstractClass{
public static void main(String [] args){
    shape c = new Circle();
    c.input();
    System.out.println("The area of circle is " + c.area());

    shape R = new Rectangle();
    R.input();
    System.out.println("The area of rectangle is " + R.area());
    
    shape T = new Triangle();
    T.input();
    System.out.println("The area of rectangle is " + T.area());
  }
}

